File: okreplay-core/src/main/java/okreplay/AbstractMessage.java
package okreplay;

import java.nio.charset.Charset;

import okhttp3.MediaType;

import static okreplay.Util.CONTENT_TYPE;
import static okreplay.Util.isNullOrEmpty;

abstract class AbstractMessage implements Message {
  private static final String CONTENT_ENCODING = "Content-Encoding";
  private static final Charset UTF_8 = Charset.forName("UTF-8");
  static final String DEFAULT_CONTENT_TYPE = "application/octet-stream";
  private static final String DEFAULT_ENCODING = "none";

  @Override public String getContentType() {
    String header = header(CONTENT_TYPE);
    if (isNullOrEmpty(header)) {
      return DEFAULT_CONTENT_TYPE;
    } else {
      return MediaType.parse(header).toString();
    }
  }

  @Override public Charset getCharset() {
    String header = header(CONTENT_TYPE);
    if (isNullOrEmpty(header)) {
      // TODO: this isn't valid for non-text data – this method should return Optional<String>
      return UTF_8;
    } else {
      return Optional.fromNullable(MediaType.parse(header).charset()).or(UTF_8);
    }
  }

  @Override public String getEncoding() {
    String header = header(CONTENT_ENCODING);
    return header == null || header.length() == 0 ? DEFAULT_ENCODING : header;
  }

  @Override public String header(String name) {
    return headers().get(name);
  }

  @Override public final String bodyAsText() {
    return new String(body(), getCharset());
  }
}

File: okreplay-core/src/main/java/okreplay/AbstractMessage.kt
package okreplay

import okhttp3.MediaType
import okreplay.Util.CONTENT_TYPE
import okreplay.Util.isNullOrEmpty
import java.nio.charset.Charset

internal abstract class AbstractMessage : Message {

  override fun getContentType(): String {
    val header = header(CONTENT_TYPE)
    return if (isNullOrEmpty(header)) {
      DEFAULT_CONTENT_TYPE
    } else {
      MediaType.parse(header!!)!!.toString()
    }
  }

  override fun getCharset(): Charset {
    val header = header(CONTENT_TYPE)
    return if (isNullOrEmpty(header)) {
      // TODO: this isn't valid for non-text data – this method should return String?
      UTF_8
    } else {
      MediaType.parse(header!!)!!.charset() ?: UTF_8
    }
  }

  override fun getEncoding(): String {
    val header = header(CONTENT_ENCODING)
    return if (header == null || header.isEmpty()) DEFAULT_ENCODING else header
  }

  override fun header(name: String): String? {
    return headers().get(name)
  }

  override fun bodyAsText(): String {
    return String(body(), charset)
  }

  companion object {
    private val CONTENT_ENCODING = "Content-Encoding"
    private val UTF_8 = Charset.forName("UTF-8")
    @JvmField val DEFAULT_CONTENT_TYPE = "application/octet-stream"
    private val DEFAULT_ENCODING = "none"
  }
}
